initrd/bin/kexec-seal-key

```diff
@@ -83,73 +87,120 @@ done
+	# now just all zeros in a normal boot
+	# PCR 5 must be all zero since no kernel modules should have
+	# been loaded during a normal boot, but might have been
+	# loaded in the recovery shell.
+	# Otherwise use the current values of the PCRs, which will be read
+	# from the TPM as part of the sealing ("X").
+	tpm sealfile2 \
+		-if "$KEY_FILE" \
+		-of "$TPM_SEALED" \
+		-pwdd "$key_password" \
+		-hk 40000000 \
+		-ix 0 X \
+		-ix 1 X \
+		-ix 2 X \
+		-ix 3 X \
+		-ix 4 0000000000000000000000000000000000000000 \
+		-ix 5 $pcr_5 \
+		-ix 6 $luks_pcr \
+		-ix 7 X
+
+	if [ $? -ne 0 ]; then
+		die "Unable to seal secret"
+	fi
 
+	shred -n 10 -z -u "$KEY_FILE" 2> /dev/null \
+		|| die "Failed to delete key file"
 
-	tpm nv_writevalue \
+	# try it without the owner password first
+	if ! tpm nv_writevalue \
 		-in $TPM_INDEX \
 		-if "$TPM_SEALED" \
-	|| die "Unable to write sealed secret to NVRAM"
+		; then
+		# to create an nvram space we need the TPM owner password
+		# and the TPM physical presence must be asserted.
+		#
+		# The permissions are 0 since there is nothing special
+		# about the sealed file
+		tpm physicalpresence -s \
+			|| warn "Warning: Unable to assert physical presence"
+
+		read -s -p "TPM Owner password: " tpm_password
+		echo
+
+		tpm nv_definespace \
+			-in $TPM_INDEX \
+			-sz $TPM_SIZE \
+			-pwdo "$tpm_password" \
+			-per 0 \
+			|| warn "Warning: Unable to define NVRAM space; trying anyway"
+
+
+		tpm nv_writevalue \
+			-in $TPM_INDEX \
+			-if "$TPM_SEALED" \
+			|| die "Unable to write sealed secret to NVRAM"
+	fi
+elif [ "$CONFIG_TPM2_TOOLS" = "y" ]; then
+	pcrf="/tmp/secret/pcrf.bin"
+	#TODO: verify logic below against tpm2
+	tpm2 pcrread -o "$pcrf" sha256:0,1,2,3
+	# pcr 4 is expected to be zero (init mode)
+	dd if=/dev/zero bs=32 count=1 >> "$pcrf"
+	if [ "$CONFIG_USB_KEYBOARD" = "y" -o -r /lib/modules/libata.ko -o -x /bin/hotp_verification ]; then
+		DEBUG "TPM2, with PCR5 involvement (additional kernel modules are loaded per board config)..."
+		# Here, we take pcr 5 into consideration if modules are expected to be measured+loaded
+		# 
+		# binary pcr dump can only go via stderr for redirection
+		# sadly this busybox has buggy support for "2>>" and "2|"
+		# so we need workaround...
+		DEBUG "tpm2 pcrread -o /dev/stderr sha256:5 2>&1 >/dev/console | cat >> "$pcrf""
+		tpm2 pcrread -o /dev/stderr sha256:5 2>&1 >/dev/console | cat >> "$pcrf"
+	else
+		DEBUG "TPM2, without PCR5 involvement. Inserting 32 zero bytes under pcrf..."
+		#no kernel modules are expected to be measured+loaded
+		dd if=/dev/zero bs=32 count=1 >> "$pcrf"
+	fi
+	# Use pcr 23 to precompute the value for pcr 6
+	tpmr extend -ix 23 -if /tmp/luksDump.txt
+	tpm2 pcrread -o /dev/stderr sha256:23 2>&1 >/dev/console | cat >> "$pcrf"
+	# goal is to validate that what is in pcr 23 is at pcr 6 at unseal
+	DEBUG "PCR23 content used to simulate PCR6 content at unseal:" 
+	DEBUG "$(pcrs)"
+	tpm2 pcrreset 23
```

Comment:
Doesn't do anything?

Reply:
Per the comment, this is computing a value for PCR 6 using PCR 23 as a temporary area.

We could probably do this on the host CPU with an equivalent computation, haven't looked into it yet.

Reply:
Sorry for having been unclear. Coming from tpm1.2, pcrfuturecalc was not modifying the PCR, just using one expected empty to produce a hash. In the code, pcr23 is expected to be "reset", which never happens (which is what I meant!)

Reply:
Nothing wrong, and ectually better, to have TPM produce the hash then to rely on local CPU. After all, this is why we use the TPM to do measurements. :)